File: backend/app/api/serializers.py
from __future__ import annotations

import json

from app.models.models import SiteSettings, TeamMember
from app.schemas import ContactLinks, MemberDetail, MemberPublic, SiteSettingsOut


def parse_json(raw: str | None, default: dict | list | None = None):
    if not raw:
        return default if default is not None else {}
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return default if default is not None else {}


def site_to_out(site: SiteSettings) -> SiteSettingsOut:
    return SiteSettingsOut(
        brand_name=site.brand_name,
        tagline=site.tagline,
        contacts=ContactLinks(
            email=site.email,
            telegram=site.telegram,
            linkedin=site.linkedin,
            github=site.github,
        ),
        home_content=parse_json(site.home_content, {}),
    )


def member_to_public(m: TeamMember) -> MemberPublic:
    return MemberPublic.model_validate(m)


def member_to_detail(m: TeamMember, include_resume: bool = True) -> MemberDetail:
    base = MemberPublic.model_validate(m)
    resume = None
    if include_resume and m.resume:
        resume = parse_json(m.resume.data_json, {})
    return MemberDetail(**base.model_dump(), resume=resume)

File: src/lib/serializers.ts
import type { TeamMember } from '@prisma/client'

type SiteRow = {
  brandName: string
  tagline: string
  email: string
  telegram: string
  linkedin: string
  github: string
  homeContent: string
}

type MemberWithResume = TeamMember & {
  resume?: { dataJson: string } | null
}

function parseJson(raw: string | null | undefined, fallback: Record<string, unknown> = {}) {
  if (!raw) return fallback
  try {
    return JSON.parse(raw) as Record<string, unknown>
  } catch {
    return fallback
  }
}

export function siteToOut(site: SiteRow) {
  return {
    brand_name: site.brandName,
    tagline: site.tagline,
    contacts: {
      email: site.email,
      telegram: site.telegram,
      linkedin: site.linkedin,
      github: site.github,
    },
    home_content: parseJson(site.homeContent, {}),
  }
}

export function memberToPublic(m: TeamMember) {
  return {
    id: m.id,
    slug: m.slug,
    name_en: m.nameEn,
    name_fa: m.nameFa,
    role_en: m.roleEn,
    role_fa: m.roleFa,
    short_bio_en: m.shortBioEn,
    short_bio_fa: m.shortBioFa,
    avatar_url: m.avatarUrl,
    email: m.email,
    github: m.github,
    linkedin: m.linkedin,
    telegram: m.telegram,
    sort_order: m.sortOrder,
    is_published: m.isPublished,
  }
}

export function memberToDetail(m: MemberWithResume) {
  return {
    ...memberToPublic(m),
    resume: m.resume ? parseJson(m.resume.dataJson, {}) : null,
  }
}

export function jsonError(detail: string, status: number) {
  return Response.json({ detail }, { status })
}
